msal/application.py

```diff
@@ -1746,9 +1720,53 @@ class PublicClientApplication(ClientApplication):  # browser app or mobile app
     def __init__(self, client_id, client_credential=None, **kwargs):
         """Same as :func:`ClientApplication.__init__`,
         except that ``client_credential`` parameter shall remain ``None``.
+
+        .. note::
+
+            You may set enable_broker_on_windows to True.
+
+            What is a broker, and why use it?
+
+            A broker is a component installed on your device.
+            Broker implicitly gives your device an identity. By using a broker,
+            your device becomes a factor that can satisfy MFA (Multi-factor authentication).
+            This factor would become mandatory
+            if a tenant's admin enables a corresponding Conditional Access (CA) policy.
+            The broker's presence allows Microsoft identity platform
+            to have higher confidence that the tokens are being issued to your device,
+            and that is more secure.
+
+            An additional benefit of broker is,
+            it runs as a long-lived process with your device's OS,
+            and maintains its own cache,
+            so that your broker-enabled apps (even a CLI)
+            could automatically SSO from a previously established signed-in session.
+
+            ADFS and B2C do not support broker.
+            MSAL will automatically fallback to use browser.
+
+            You shall only enable broker when your app:
+
+            1. is running on supported platforms,
+               and already registered their corresponding redirect_uri
+
+               * ``ms-appx-web://Microsoft.AAD.BrokerPlugin/your_client_id``
+                 if your app is expected to run on Windows 10+
+
+            2. installed broker dependency,
+               e.g. ``pip install msal[broker]>=1.25,<2``.
+
+            3. tested with ``acquire_token_interactive()`` and ``acquire_token_silent()``.
+
+        :param boolean enable_broker_on_windows:
+            This setting is only effective if your app is running on Windows 10+.
+            This parameter defaults to None, which means MSAL will not utilize a broker.
         """
         if client_credential is not None:
             raise ValueError("Public Client should not possess credentials")
+        # Using kwargs notation for now. We will switch to keyword-only arguments.
+        enable_broker_on_windows = kwargs.pop("enable_broker_on_windows", False)
```

Comment:
Unlike `allow_broker` which is supported by both `PublicClientApplication` and `ClientApplication`, `enable_broker_on_windows` is now only supported by `PublicClientApplication`. Doesn't https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/559 stand anymore?

Reply:
Just because `allow_broker` is deprecated, #559 should no longer be considered a reasonable design.

Using the old `allow_broker` on a `ConfidentialClientApplication` fails on the application level:

https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/9ad3fd7565c91667e111063ee560e1cbdbe53b41/msal/application.py#L570-L571

But now using the new `enable_broker_on_windows` on a `ConfidentialClientApplication` fails on Python level:

```py
  File "d:\cli\azure-cli\src\azure-cli-core\azure\cli\core\auth\msal_authentication.py", line 132, in __init__
    super().__init__(service_principal_auth.client_id, client_credential=client_credential, **kwargs)
TypeError: ClientApplication.__init__() got an unexpected keyword argument 'enable_broker_on_windows'
```

The reason I would like to confirm the new design is because Azure CLI has an open PR https://github.com/Azure/azure-cli/pull/26482 to address the using `allow_broker` on `ConfidentialClientApplication` issue. If using `enable_broker_on_windows` is limited to `PublicClientApplication`, I will make the corresponding change in Azure CLI.

Reply:
> Just because `allow_broker` is deprecated, #559 should no longer be considered a reasonable design.

Why is #559 even relevant here? It was just a one-line document change which does not change any behavior.

Let's get back to the topic of this PR. This PR does not really change the behavior, either, because the intention of rejecting `ConfidentialClientAppication(..., broker=True)` remains the same. Your Azure CLI PR of adopting the new parameter also looks good to me.

Are we on the same page that this PR itself can be merged in, while we will continue working on the username password flow and device code flow issues separately?

Reply:
note: broker is a public client application concept only